enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,929 @@
+
+#### Backup contents
+
+Entire MicroShift data directory will be backed up, this includes etcd database,
+certificates, and kubeconfigs.
+
+- Copying entire etcd working directory will preserve history and other metadata
+  that would have been lost when using etcd snapshots.
+- Not regenerating certificates on each upgrade will keep them valid. It also
+  means that kubeconfigs will continue to work as opposed to needing to obtain
+  them again.
+
+### Backup retention
+
+Because a user may stage multiple ostree deployments on a host and boot them in
+any order, we will keep multiple backups to ensure that we can roll back to a
+state compatible with any ostree deployment
+
+Because we want to minimize the impact of backups on storage requirements,
+we will keep only 1 backup per ostree deployment.
+
+### Migrating data between versions
+
+Term "data migration" is referring to process of transforming data from one
+schema version to another.
+
+It needs to be aware of following data types:
+- Kubernetes storage migration (e.g. from `v1beta1` to `v1`)
+  - MicroShift will reuse
+    [Cluster Kube Storage Version Migrator Operator](https://github.com/openshift/cluster-kube-storage-version-migrator-operator)
+- etcd schema (although it's unlikely in near future)
+- Internal MicroShift-specific data
+
+Data migration will be attempted only with etcd and kube-apiserver running, with
+any other components or workloads not running to keep them from interfering.
+
+Because we want to maintain upgrade expectations with Kubernetes and OpenShift,
+data migration will only be supported from Y to Y+1 version (although it might
+change if upstream components begin to support greater version skews).
+
+Because we cannot guarantee that the data formats between Y versions are
+compatible after an upgrade, going from Y to Y-1 will only be possible when
+rolling back to commit (with older MicroShift) with existing backup.
+
+If MicroShift data directory does not contain version information, it will be
+assumed that it was created by MicroShift 4.13 and tested with Y-stream skew
+rule.
+
+Because we may need to counteract regressions within a Y version and because
+MicroShift is minimal version of OpenShift (so risk of Z-stream incompatibilities
+is greatly reduced), at the time of writing the enhancement,
+switching between different Z streams will be possible regardless of the
+direction (older to newer, newer to older) and anything making this impossible
+(like breaking change or backport) should be documented.
+
+Because we may need to block certain upgrade sequences, similar to OpenShift's
+upgrade graph, but we cannot ensure access to that upgrade graph from edge
+systems and we cannot prevent an attempted upgrade via a new ostree deployment,
+we will incorporate a mechanism to block specific upgrades by listing version
+numbers _from which_ a new version cannot be upgraded (X.Y+1.Z may include
+X.Y.Z in its "block" list).
```

Comment:
"Building MicroShift upgrade graph" is currently listed as a non-goal, so you might want to drop this paragraph?  If you keep it, I'm unclear about how this data will be managed.  For traditional OpenShift, there's [this process][1] for flagging issues with update-regression risks, allowing us to declare risks when updating into already-released versions (e.g. openshift/cincinnati-graph-data#3576).  Sometimes it takes more than a few days to get the fix in position, and we end up extending the known risk to later releases (e.g. openshift/cincinnati-graph-data#3611).  But often we fix the regression in the next release, and do not need to extend the risk declaration.

I'm not all that familiar with `block`, but I'm curious about whether it can only be used for regressions that you know about at MicroShift-RPM-build-time.  And if you know about those regressions at build time, it's not clear to me whether it's worth cutting a MicroShift release at all.

Or is this for things like "we're still working some kinks out of X.(Y-1) updates, so push folks through X.(Y-1).Z+ before they head to X.Y.Z', like openshift/cincinnati-graph-data#3581?

Anyhow, lots of options, and I'm interested in whether we keep the `block` reference in this enhancement, and if we do, which teams are maintaining that metadata and what the process might look like.

[1]: https://github.com/openshift/enhancements/blob/ae7d49d2c1ab51886417e89353b754f873351100/enhancements/update/update-blocker-lifecycle/README.md

Reply:
Actually I haven't spent much time on that part of the updateability, but for me it's more like:
> Or is this for things like "we're still working some kinks out of X.(Y-1) updates, so push folks through X.(Y-1).Z+ before they head to X.Y.Z', like https://github.com/openshift/cincinnati-graph-data/pull/3581?

We cannot amend releases retroactively, so only option is to restrict that you can upgrade to 4.14.20 from 4.14.10 and newer, or maybe to upgrade to 4.15.10, you need at least 4.14.10.
Given that we need to support X.Y -> X.Y+1, it seems to give us power only in terms of Z-streams, so it might be limited compared to what OpenShift does, but it would be good to have this in place from the start.

For where that data is stored - since this is intended to be embedded into a binary, I imagined `microshift` repository. We're yet to define the process (probably another enhancement), but we'll need to carefully navigate between areas like ushift code, ushift rebase procedure, ART build system, and QE.